File: wayland/src/main/java/org/westmalle/wayland/core/Region.java
//Copyright 2015 Erik De Rijcke
//
//Licensed under the Apache License,Version2.0(the"License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing,software
//distributed under the License is distributed on an"AS IS"BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
package org.westmalle.wayland.core;


import javax.annotation.Nonnull;
import java.util.List;

public interface Region {

    @Nonnull
    List<Rectangle> asList();

    @Nonnull
    Region add(@Nonnull Rectangle rectangle);

    @Nonnull
    Region subtract(@Nonnull Rectangle rectangle);

    boolean contains(@Nonnull Point point);

    boolean contains(@Nonnull Rectangle clipping,
                     @Nonnull Point point);
}

File: compositor/src/main/java/org/westford/compositor/core/Region.kt
/*
 * Westford Wayland Compositor.
 * Copyright (C) 2016  Erik De Rijcke
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.westford.compositor.core

interface Region {

    fun asList(): List<Rectangle>

    fun add(rectangle: Rectangle)

    fun subtract(rectangle: Rectangle)

    operator fun contains(point: Point): Boolean

    fun contains(clipping: Rectangle,
                 point: Point): Boolean

    operator fun contains(rectangle: Rectangle): Boolean

    fun intersect(rectangle: Rectangle): Region

    fun copy(): Region

    fun isEmpty(): Boolean
}

